compiler/src/iree/compiler/Dialect/Flow/Transforms/CollapseDimensions.cpp

```diff
@@ -416,44 +571,200 @@ hoistTensorReshapesOutOfDispatchRegion(RewriterBase &rewriter,
   return newDispatchOp;
 }
 
-/// Traverses DispatchRegionOps to find linalg genericOps and collapses
-/// dimensions without modifying operands with producers
-static bool collapseDimensions(IRRewriter &rewriter,
-                               DispatchRegionOp &regionOp) {
-  // Step 1. Find the root linalg.generic Op
-  std::optional<linalg::GenericOp> genericOp = findRootGenericOp(regionOp);
-  if (!genericOp.has_value())
-    return false;
+//===---------------------------------------------------------------------===//
+// Collapse shape propagation
+//===---------------------------------------------------------------------===//
+
+// For each consumer, use it's producers to constrain which dimensions it will
+// collapse. `slice` is expected to be topologically sorted (getBackwardSlice
+// does this automatically).
+static void updateConsumersFromProducers(
+    ArrayRef<Operation *> slice,
+    llvm::MapVector<linalg::GenericOp, CollapseInfo> &opMap) {
+  for (auto op : slice) {
+    auto genericOp = cast<linalg::GenericOp>(op);
+    CollapseInfo &consumerInfo = opMap[genericOp];
+
+    for (auto operand : genericOp.getDpsInputOperands()) {
+      auto definingOp = operand->get().getDefiningOp();
+      if (!definingOp || isNonNullAndOutsideDispatch(definingOp)) {
+        continue;
+      }
+
+      // Track the dimensions that are not collapsable by this current op. Note:
+      // the dims are relative to the consumers iteration space, not the
+      // producers. This allows for an easy set_subtract.
+
+      CollapseInfo::CollapsableLoopsSet producerUncollapsable;
+      for (auto expr : genericOp.getMatchingIndexingMap(operand).getResults()) {
+        producerUncollapsable.insert(cast<AffineDimExpr>(expr).getPosition());
+      }
+
+      auto producerOp = dyn_cast<linalg::GenericOp>(definingOp);
+      FailureOr<AffineMap> mapping =
+          getProducerLoopToConsumerLoopsMap(*operand);
+
+      // If the producer is not a generic or there is no mapping, the tensor is
+      // not collapsable.
+      if (!producerOp || failed(mapping)) {
+        consumerInfo.updateCollapseViaSubtract(producerUncollapsable);
+        continue;
+      }
 
-  // Step 2. Check whether it is possible to collapse
-  if (!isEligibleForCollapse(genericOp.value()))
+      CollapseInfo &producerInfo = opMap[producerOp];
+      FailureOr<CollapseInfo::CollapsableLoopsSet> producerCollapsable =
+          producerInfo.getTransformedCollapsableLoops(mapping.value());
+      if (!failed(producerCollapsable)) {
+        producerUncollapsable.set_subtract(producerCollapsable.value());
+      }
+
+      consumerInfo.updateCollapseViaSubtract(producerUncollapsable);
+    }
+  }
+}
+
+// For each producer, use it's consumers to constrain which dimensions it will
+// collapse. `slice` is expected to be topologically sorted (getBackwardSlice
```

Comment:
THis uses forward slice?

Reply:
No, I just use `llvm::reverse` on the slice created by getBackwardSlice so that each ops producer has been updated before before updating said op.